### DISCUSSION TITLE
Update to the app.asar file in Rekordbox v6.6.5 breaks database key extraction

### DISCUSSION THREAD

**Discussion by dylanljones** [dylanljones/pyrekordbox#97] [Rekordbox]
Pioneer changed the ``app.asar`` file contents in Rekordbox version ``6.6.5``. 
The encryption password of the database key is no longer stored in plain text, which breaks the database unlocking.

Previously, the ``app.asar`` file contained JS files in plain text. 
Since update 6.6.5 the JS files are now stored in a compiled format (.jsc).
The password should still be somewhere in the content of the file (``jsc/controllers/auth_manager.jsc``), but it can no longer be extracted easily.

Please feel free to join the discussion if you have any ideas!

**Environment**
- Rekordbox version: >=6.6.5

**Comment by morenoh149** [comment]
can a copy of the .jsc be shared in this thread?

**Comment by dylanljones** [comment]
I am not sure if it is a good idea to share parts of the source code of Pioneer here, but i can send you an Email if you want to have a look at the ``auth_manager.jsc`` file. 
If you have Rekordbox 6.6.5 installed, it is also really easy to extract the file from the asar archive yourself. I made a little tool to print out or extract single files form an archive, if you are interested, it is [here](https://github.com/dylanljones/asarlib). The ``app.asar`` file is located in RBv6 program data directory in ``.../rekordbox 6.6.5/rekordboxAgent-win32-x64/resources``

**Comment by cheung31** [comment]
@dylanljones would `asarlib` allow one to inspect the `auth_manager.jsc` for the password?

**Comment by dylanljones** [comment]
The `asarlib` package only extracts the contents of the archive, so it is not possible to read the password. The `auth_manager.jsc` is not stored in plain text, but the password should still be stored in the file somewhere, although scrambled.

**Comment by niderhoff** [comment]
Hey guys, just want to chime in and ask if I understand the problem correctly.

I am using pyrekordbox master installed from git, Rekordbox v 6.6.5 and I was trying to run the rekordbox 6 example code from the readme (https://github.com/dylanljones/pyrekordbox#rekordbox-6-database). I am directly using the example code, which I copied directly from the readme:

```python
from pyrekordbox import Rekordbox6Database

db = Rekordbox6Database()

for content in db.content():
	print(content.Title, content.Artist.Name)
```

However, it would always give me the following error, which was a bit confusing to me at first glance: "WARNING:root:No Rekordbox 5 folder found in /Applications". Then it would crash.

Then I turned on the debugger and found that the problem must be inside the `_get_rb6_config` function which is invoked from `config.py` during import. It in turn invokes the `_get_rb_config` function, which I assume is for rekordbox 5, hence the confusing warning.

Apart from that, the function `_get_rb6_config` correctly identifies my `master.db` location. 

The following procedures of said function include
```python
# Read password from app.asar, see
# https://www.reddit.com/r/Rekordbox/comments/qou6nm/key_to_open_masterdb_file/

asar_data = read_rekordbox6_asar(conf["install_dir"])
match = re.search('pass: ".(.*?)"', asar_data).group(0)
```

It looks like something is read from the asar file but the regex does not match `'pass:'` anywhere in the asar file. This results in a `KeyError`, because `group(0)` tries to index in an empty list and the error is not handled. 

Question 1: Am I right in assuming that this line (i.e. the regex matching part) needs to be swapped out for some clever parsing of the asar_data which can read the obfuscated password?

You mention that 

> The asarlib package only extracts the contents of the archive, so it is not possible to read the password. The auth_manager.jsc is not stored in plain text, but the password should still be stored in the file somewhere, although scrambled.


After some googling I found that `.jsc` is most likely the nodejs equivalent of something like bytecode. So it would need to be de-compiled to be read properly, and according to my research that is quite a dead end, since there have been no public attempts to de-compile .jsc files generated with node > 8.4 in the past 2 years or so.

Question 2: You think this might be true or is it some other kind of obfuscation in place here? In general I am not very experienced in reverse engineering... any ideas how we could go about it?

I tried to extract the contents of the asar using `asarlib` but it failed because of some KeyError on (self.header["files"]) at some point in the extraction loop, so I could not verify or further analyze the structure or contents of the asar file.

**Comment by 11ib** [comment]
Does anyone have the path to the `auth_manager.jsc` file? I couldn't find it on a fresh install of Rekordbox (6.6.8) on Windows.

**Comment by dylanljones** [comment]
Hey everybody, sorry for the late response!

@niderhoff, regarding question 1, you are totally right. Before we could just do a regex search, but ever since update 6.6.5 all the relevant files in the ``app.asar`` archive are scrambled. I also looked for a clever way to hack the ``.jsc`` file, but with no result. 
So I think it really is a "compiled" version of a JavaScript file, but I have no clue how one could de-compile it. This should also answer your second question... But I still have some hope left that it can be reverse engineered somehow, however I am also not too experienced in that field.
The warning due to the missing Rekordbox v5 installation was a bad design decision on my part. Didn't notice it since I still have v5 installed, but I will fix that. Thanks for letting me know:) Also sorry to hear that ``asarlib`` failed, to be honest I only tested it with the ``app.asar`` from my Rekordbox installation.

@11ib, the ``auth_manager.jsc`` is stored in the ``app.asar`` file, which is a kind of archive that contains files and directories. There are tools out there to extract the contents (like the one I wrote, which is pretty buggy). The path for the archive on Windows is something like
````
C:\Program Files\Pioneer\rekordbox 6.5.3\rekordboxAgent-win32-x64\resources\app.asar
````

Hope that answers a few question, I wish I knew more. And thank you for your interest!

**Comment by niderhoff** [comment]
Apparently .jsc is compiled v8 bytecode. We can disassemble v8 bytecode, but it will not bring back javascript, but instead human readable v8 instructions: https://github.com/noelex/v8dasm. Maybe then we can look through it look for something that might look like the password.

Another route would be if you knew the format of the password (assuming it has not changed since rekordbox 6.4, you could search in the hex dump for a string which has roughly the same length and type of characters.)

**Comment by DanielMS93** [comment]
Any progress on this guys? i'm hitting the same issue i believe

<img width="581" alt="image" src="https://github.com/dylanljones/pyrekordbox/assets/44974943/8826a5dd-b98d-4215-8e71-071c7b59ca5e">

**Comment by dylanljones** [comment]
Hey, sorry for the inactivity, have a lot of other stuff going on right now.
Sadly no progress so far, but I am looking into a few things! The key is still the same as in earlier Rekordbox versions, but i would **really** like to avoid just hard-coding it for legal reasons (as discussed at the end of [this](https://rekord.cloud/blog/technical-inspection-of-rekordbox-6-and-its-new-internals) blog of rekordcloud).

**Comment by niderhoff** [comment]
Hey Dylan I might have a idea, can I contact you somehow?

**Comment by dylanljones** [comment]
Sorry, just realized my Email wasn't showing up. It should be on my profile now. You can contact me there!

**Comment by DanielMS93** [comment]
Could we use blackbox for credentials?

**Comment by puhitaku** [comment]
The disassembly of a simple hello world shows that the string "Hello World!" is stored in a "Constant pool" as-is, like other usual executables.

```
$ cat app.js
function foo() {
    console.log("Hello World!")
}

foo()

$ node --print-bytecode --print-bytecode-filter=foo app.js
[generated bytecode for function: foo (0x1f66e1498a39 <SharedFunctionInfo foo>)]
Bytecode length: 19
Parameter count 1
Register count 3
Frame size 24
OSR urgency: 0
Bytecode age: 0
   21 S> 0x1f66e1499ac0 @    0 : 21 00 00          LdaGlobal [0], [0]
         0x1f66e1499ac3 @    3 : c3                Star1
   29 E> 0x1f66e1499ac4 @    4 : 2d f9 01 02       GetNamedProperty r1, [1], [2]
         0x1f66e1499ac8 @    8 : c4                Star0
         0x1f66e1499ac9 @    9 : 13 02             LdaConstant [2]
         0x1f66e1499acb @   11 : c2                Star2
   29 E> 0x1f66e1499acc @   12 : 5e fa f9 f8 04    CallProperty1 r0, r1, r2, [4]
         0x1f66e1499ad1 @   17 : 0e                LdaUndefined
   49 S> 0x1f66e1499ad2 @   18 : a9                Return
Constant pool (size = 3)
0x1f66e1499a61: [FixedArray] in OldSpace
 - map: 0x12fa427812e1 <Map>
 - length: 3
           0: 0x12fa42784a01 <String[7]: #console>
           1: 0x32d47820c731 <String[3]: #log>
           2: 0x1f66e14999f9 <String[12]: #Hello World!>
Handler Table (size = 0)
Source Position Table (size = 10)
0x1f66e1499ad9 <ByteArray[10]>
Hello World!
```

I compiled the app.js into jsc and the constant is still intact in the binary.

```sh
$ npm install -g bytenode
$ bytenode -n -c app.js
$ grep "Hello World!" app.jsc
Binary file app.jsc matches
```

As @dylanljones pointed out, the key is the same as in the previous version, and the key may be stored in the jsc intact if it is not obfuscated in a way we don't expect.

Expecting that the key is a string literal, this gives us a way to "match" the key string that fulfills the known format with surrounding fingerprints of constants. Embedding the match condition (not the key itself) in this repository will not cause any legal concerns.

How about this idea?

**Comment by puhitaku** [comment]
I wish I could get Rekordbox older than 6.6.5 ...

**Comment by dylanljones** [comment]
Hi @puhitaku,

thanks for joining in on the discussion! Sadly, the key is not a string literal and is obfuscated in some way. I already tried to match the known key - without any success...
Do you know if there is an option for that when generating the .jsc? If so, there might be a way to reverse the obfuscation.

@DanielMS93, thank you for the hint! As far as i understand, blackbox just encrypts the 'shipped' keys, right? I did not spend a lot of time looking into it yet, but it seems we still would have to ship the (encrypted) key, right? But i will look a bit more into blackbox:)

**Comment by dylanljones** [comment]
Maybe we should use some workaround until we have figured out how to de-obfuscate the .jsc files (as some other projects do, see issue #77)

**Comment by niderhoff** [comment]
One could pass the key to the db handler. This is how it would look:

```
from pyrekordbox import Rekordbox6Database

db = Rekordbox6Database(key="<insert key here>")

for content in db.get_content():
    print(content.Title, content.Artist.Name)

playlist = db.get_playlist()[0]
for song in playlist.Songs:
    content = song.Content
    print(content.Title, content.Artist.Name)
```

if you had an old installation, `rb.cache` file, or the key noted down, you can just use it like thit.

If not one could go to #77 where someone linked to another repository where the developer(s) have embedded the key into the code.

**Comment by dylanljones** [comment]
The key of the Rekordbox v6 database can now be downloaded and cached from external sources (mainly the projects in #77) using the command line interface:
````shell
python -m pyrekordbox download-key
````

This is still just a workaround, it still would be nice to find a way to extract the key from the `.jsc`-files, especially if Pioneer decides to change the key (maybe in Rekordbox v7?). In the mean time this fix should get the project running for everyone:)

If anyone finds some more sources for the key, feel free to add them to the CLI or post the links here in case the other projects remove the key!

**Comment by dylanljones** [comment]
I moved this thread to the discussions since it is not an issue at the moment, but any new ideas are welcome!

**Comment by Bide-UK** [comment]
Wow I didn't realise there was this community of people also trying to make rekordbox actually worth using. 

I created this script to pull out the key in rage after I realised I'd have to manually organise my 15 years of music on my new machine.
This in theory should be pretty resilient against future changes / obfuscation attempts. But it does require SIP to be deactivated which is not ideal. 

https://github.com/Bide-UK/rekordlocksmith

**Reply by dylanljones** [reply]
Hey @Bide-UK , that's exactly the goal of the project:D

The script of you is very nice, if its okay with you i will add parts of your README and a link of your repo to the docs. 
If we can figure out a similar method for Windows we should be clear for future versions of Rekordbox.
Thank you for commenting and making the script!

**Reply by Bide-UK** [reply]
Please do! The logic should be somewhat similar but I don't actually have a windows machine, I'll keep you posted if I make any changes.

**Comment by dylanljones** [comment]
Just found [frida](https://github.com/frida/frida) for code injections, maybe that's also an option for extracting the key. See for example [this repo](https://github.com/liamcottle/pioneer-rekordbox-database-encryption). Will look into this!

**Reply by Bide-UK** [reply]
I had a POC in Frida but because I'm on M1 and Pioneer CBA to recompile rekordbox for ARM due to the deemix dependancies in the binary I was unable to get it to work. Let me fire up my old machine and see what I can do.

**Reply by dylanljones** [reply]
Got it to work via a Python script!


<details><summary>Code</summary>

````python
import time
import frida

SCRIPT = """
var sqlite3_key = Module.findExportByName(null, 'sqlite3_key');

Interceptor.attach(sqlite3_key, {
    onEnter: function(args) {
        var size = args[2].toInt32();
        var key = args[1].readUtf8String(size);
        send('sqlite3_key: ' + key);
    }
});
"""

class KeyExtractor:

    def __init__(self, rekordbox_executable):
        self.executable = rekordbox_executable
        self.key = ""

    def on_message(self, message, data):
        payload = message["payload"]
        if payload.startswith("sqlite3_key"):
            self.key = payload.split(": ")[1]

    def run(self):
        # Spawn Rekordbox process and attach to it
        pid = frida.spawn(self.executable)
        frida.resume(pid)
        session = frida.attach(pid)
        script = session.create_script(SCRIPT)
        script.on("message", self.on_message)
        script.load()
        # Wait for key to be extracted
        while not self.key:
            time.sleep(1)
        # Kill Rekordbox process
        frida.kill(pid)

        return self.key

extractor = KeyExtractor("path/to/Rekordbox_executable")
key = extractor.run()
````

</details>

This will open the Rekordbox application and extract the key when the database is opened for the first time. Can anyone check f this works on macs? Whats the issue on apple silicon? Would be great to get this working on all platforms.

**Comment by dylanljones** [comment]
I think at this point we have enough options, closing for now:)
